adsmp/solr_updater.py

```diff
@@ -310,44 +366,83 @@ def transform_json_record(db_record):
+        links_data = db_record["bib_data"].get("links_data", None)
         if links_data:
             try:
                 links_data = json.loads(links_data[0])
-                if 'property' not in out:
-                    out['property'] = []
-                if links_data.get('access', None) == 'open':
-                    out['property'].extend(['ESOURCE',
-                                            'ARTICLE',
-                                            'NOT REFEREED',
-                                            'EPRINT_OPENACCESS',
-                                            'OPENACCESS'])
-                    if 'esources' not in out:
-                        out['esources'] = []
-                    out['esources'].extend(['EPRINT_HTML', 'EPRINT_PDF'])
+                if "property" not in out:
+                    out["property"] = []
+                if links_data.get("access", None) == "open":
+                    out["property"].extend(
+                        [
+                            "ESOURCE",
+                            "ARTICLE",
+                            "NOT REFEREED",
+                            "EPRINT_OPENACCESS",
+                            "OPENACCESS",
+                        ]
+                    )
+                    if "esources" not in out:
+                        out["esources"] = []
+                    out["esources"].extend(["EPRINT_HTML", "EPRINT_PDF"])
             except (KeyError, ValueError):
-                    # here if record holds unexpected value
-                    logger.error('invalid value in bib data, bibcode = {}, type = {}, value = {}'.format(db_record['bibcode'], type(links_data), links_data))
+                # here if record holds unexpected value
+                logger.error(
+                    "invalid value in bib data, bibcode = {}, type = {}, value = {}".format(
+                        db_record["bibcode"], type(links_data), links_data
+                    )
+                )
+
+    # Read-in names of fields to check for solr "has:" field
+    hasfields = sorted(config.get("HAS_FIELDS", []))
+
+    # populate "has:" field with fields that exist for a particular record
+    has = []
+    for field in hasfields:
+        if out.get(field, ""):
+            # if field is not empty, check if at least one character is alphanumeric
+            # this is done to not count fields where blank entries can be ['-',...]
+
+            # if field does not have a string or list, make it a list of strings so it is iterable
+            if not (isinstance(out[field], list) or isinstance(out[field], str)):
+                out_field = [str(out[field])]
+            else:
+                out_field = out[field]
+            # iterate through each character of each element in the field to check for at least one alphanumeric character
+            if any([any(char.isalnum() for char in element) for element in out_field]):
```

Comment:
I'm worried this will take a while, especially for a big field (e.g. the body of a paper, or the author list of a 1000-author paper). What about replacing `for char in element` with `for char in set(element)`? That'll remove the duplicate characters in a given field. You could optimize further by turning an array into a set of characters - something like this, a few lines above:
```
if isinstance(out[field], list):
     out_field = set(''.out[field])
```